pinot-core/src/main/java/org/apache/pinot/core/operator/filter/RangeIndexBasedFilterOperator.java

```diff
@@ -131,27 +139,75 @@ public String toExplainString() {
   }
 
   interface RangeEvaluator {
+
+    Set<FieldSpec.DataType> SUPPORTED_RAW_DATA_TYPES = EnumSet.of(FieldSpec.DataType.INT,
+            FieldSpec.DataType.LONG, FieldSpec.DataType.FLOAT, FieldSpec.DataType.DOUBLE);
+
+    static boolean canEvaluate(PredicateEvaluator predicateEvaluator, DataSource dataSource) {
+      if (dataSource.getRangeIndex() != null) {
+        boolean datatypeSupported = dataSource.getRangeIndex().isExact()
+                && (predicateEvaluator.isDictionaryBased()
+                || SUPPORTED_RAW_DATA_TYPES.contains(predicateEvaluator.getDataType()));
+        switch (predicateEvaluator.getPredicateType()) {
+          case EQ:
+            return datatypeSupported && dataSource.getRangeIndex().isExact()
+                    && predicateEvaluator instanceof EqualsPredicateEvaluatorFactory.EqualsPredicateEvaluator;
+          case RANGE:
+            return datatypeSupported && (predicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof IntRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof LongRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof FloatRawValueBasedRangePredicateEvaluator
+                    || predicateEvaluator instanceof DoubleRawValueBasedRangePredicateEvaluator);
+          default:
+        }
+      }
+      return false;
+    }
+
     static RangeEvaluator of(RangeIndexReader<ImmutableRoaringBitmap> rangeIndexReader,
         PredicateEvaluator predicateEvaluator) {
-      if (predicateEvaluator instanceof SortedDictionaryBasedRangePredicateEvaluator) {
-        return new IntRangeEvaluator(rangeIndexReader,
-            (SortedDictionaryBasedRangePredicateEvaluator) predicateEvaluator);
+      if (predicateEvaluator.isDictionaryBased()) {
+        if (predicateEvaluator instanceof EqualsPredicateEvaluatorFactory.EqualsPredicateEvaluator) {
+          return new IntPointEvaluator(rangeIndexReader,
+                  (EqualsPredicateEvaluatorFactory.EqualsPredicateEvaluator) predicateEvaluator);
```

Comment:
Can we follow the same way as how we extract the predicate value from the range predicate evaluator: make the child class public, and add a method getValue() or getDictId(). I don't see much value added by this new interface, and tend to keep the handling consistent

Reply:
I added the interface so that the operator knows less about the predicate evaluators to reduce coupling - I think this approach is better and could be retrofitted onto the range predicate evaluators. If it's a dealbreaker, I'll change it, but I think this is better.

Reply:
On second thoughts, I don't think we're far enough apart to go back and forth about this, so I've just made the suggested change.

Reply:
However, I have had to move the inner classes up to top level to make the class names shorter, otherwise we run into a particular case where the automatic formatting produces lines longer than 120 characters and gets rejected by checkstyle 🤦🏻